core/services/mongodb-remote/src/main/java/com/mongodb/stitch/core/services/mongodb/remote/sync/internal/DataSynchronizer.java

```diff
@@ -969,435 +895,571 @@ private void syncLocalToRemote() {
-                committedEvent.getNamespace(),
-                committedEvent.getDocumentKey(),
-                committedEvent.getUpdateDescription(),
-                false));
-
-
+          if (setPendingWritesComplete) {
             docConfig.setPendingWritesComplete(nextVersion);
-            if (committedEvent.getOperationType() != OperationType.DELETE) {
-              syncWriteModelContainer.configs.add(
-                  new ReplaceOneModel<>(CoreDocumentSynchronizationConfig.getDocFilter(
-                          nsConfig.getNamespace(), docConfig.getDocumentId()),
-                        docConfig));
-            }
-          } else {
-            // v. Otherwise, invoke the collection-level conflict handler with the local change
-            // event and the remote change event (synthesized by doing a lookup of the document or
-            // sourced from the listener)
-            final ChangeEvent<BsonDocument> remoteChangeEvent;
-            if (!remoteDocumentFetched) {
-              remoteChangeEvent =
-                  getSynthesizedRemoteChangeEventForDocument(remoteColl, docConfig.getDocumentId());
-            } else {
-              remoteChangeEvent =
-                  getSynthesizedRemoteChangeEventForDocument(
-                      remoteColl.getNamespace(),
-                      docConfig.getDocumentId(),
-                      remoteDocument);
-            }
-            syncWriteModelContainer.merge(resolveConflict(
-                nsConfig.getNamespace(),
-                docConfig,
-                remoteChangeEvent));
+          }
+
+          if (action != null) { // for now, since we do remote ops inline
+            localSyncWriteModelContainer.merge(
+                performAction(nsConfig, docConfig, remoteChangeEvent, action, message,
+                    SyncMessage.L2R_METHOD, syncException)
+            );
           }
         }
-        syncWriteModelContainer.commitAndClear(getLocalCollection(
-            nsConfig.getNamespace()),
-            getUndoCollection(nsConfig.getNamespace()),
-            nsConfig.getDocsColl());
+        localSyncWriteModelContainer.commitAndClear();
       } finally {
         nsConfig.getLock().writeLock().unlock();
         streamerLock.writeLock().unlock();
       }
     }
 
-    logger.info(String.format(
-        Locale.US,
-        "t='%d': syncLocalToRemote END",
-        logicalT));
+    if (logger.isInfoEnabled()) {
+      logger.info(String.format(
+          Locale.US,
+          "t='%d': syncLocalToRemote END",
+          logicalT));
+    }
 
     // 3. If there are still local writes pending for the document, it will go through the L2R
     //    phase on a subsequent pass and try to commit changes again.
+
   }
 
-  private void emitError(final CoreDocumentSynchronizationConfig docConfig,
-                         final String msg) {
-    this.emitError(docConfig, msg, null);
+  private LocalSyncWriteModelContainer performAction(
```

Comment:
[!important] There's a strong argument to be made that all of these parameters should be a part of a struct– maybe `SyncActionMessage`. A container would also make the flow within each method easier to follow.

[nit] This method doesn't necessarily perform an action– that verbiage can be confusing since it returns a `LocalSyncWriteModelContainer`. It should either have a more semantically appropriate name (`applyAction`, `processAction`) or have a docString explaining what's going on.

Reply:
frankly I was never on-board with the Effective Java "use no more than 2, sometimes 3" parameters nonsense - creating extra unneeded objects to pass around adds more gc overhead which can impact your performance in unpredictable ways. I'm not married to it being either way I just don't tend to like to create random data classes that get used in exactly one spot

nit fixed

Reply:
changed to enqueueAction